### DISCUSSION TITLE
Using a Bridge and Performance Expectations

### DISCUSSION THREAD

**Discussion by helo-head** [HomeSpan/HomeSpan#142] [Q&A]
I was rencently troubleshooting some performance issues on my M5Stickc+ hardware and found some surpising results I thought I would query the group about ... My configuration is a bridge with two devices (fan & temp). My initial issue was a video flickering problem that turned out to be related to the temperature loop. Almost all of the temperature code was straight from the example and in particular used the timeVal()>5000 constraint. For some strange reason for each loop where the timeVal constraint was **not** met, I would see significant flickering of the LCD. There was no other code outside of the constraint so effectively it was doing nothing in each loop for 5 seconds. Changing the time changed the flicker interval. After a few hours of debugging I opted to remove the temperature device and move the temperature code to the main loop since it did not need to be exposed to Homekit. For what ever reason this eliminated the flicker issue. My second issue was related to constant timeouts. After moving to a bridge envionment I was seeing typical response time at the clients drop from 1-2 seconds down to 15-30 seconds with a significant number of timeouts. After removing most of my code and getting no closer to solving the issue, I finally opted to remove the bridge service. Again for what ever reason this seemed to solve my performance problem. Im now down to ~1 second response times. 

Im a bit confused on why I am seeing such a big difference between bridge and non bridge configurations. Any thoughts would be appreciated. As it stands today my current project can live without the bridge, however my next project will probably benefit from it so I would love to better understand what is expected with regards to bridge performance.

**Comment by HomeSpan** [comment]
Can you check the HomeSpan log files on the Serial Monitor (you'll need to set the log level to 2 in order to see all the detailed messages)?  This should show all request/responses from the Home App and might help pin down whether the issue is within HomeSpan or elsewhere.  Also, do you have a lot of devices connected to HomeSpan - typing 's' in the Serial Monitor should show all current connections.  I have sometimes found that a request I make through my iPhone is routed through one of my Home Pods before getting to HomeSpan, which definitely slows things down.

I may explore adding a "keep-alive" feature to HomeSpan that automatically sends out notification events periodically to try to keep HomeKit refreshed, though HAP guidelines do not suggest anything like this is needed.

**Comment by helo-head** [comment]
Answer to your questions as well as further information.

1. Number of clients: between 2 and 7  - performance issues do not seem related to number of connections, I see performance issues even with only two connections.
2, Hubs - HomePod mini, ATV-4, ATV-4K with the HomePod mini currently acting as master.  
2. Most status updates are thru HomePod. I assume its because its currently the master with both ATV’s being standby. 
3. Problematic bridge configuration: one bridge, one fan accessory and one temperature accessory. 
3. Ive noted two issues when using a bridge setup 
	A. Flicker of LCD directly related to loop delay in temperature accessory. 
	B. Performance delays lead to many timeouts. Its not unusual to see a delay of 5-10 seconds for the initial pair-verify commands to show up in the log. Usually there will be 10 to 15 commands bundled including the initial pair-verify


Why 90% of the status updates come thru the HomePod I cannot explain other than its the hub. It may be related to the fact that all Homekit devices (including hubs) reside on a vlan while all clients reside on main lan. 


The only differences between a the code that performs well and the code that has performance issues are the following:

The problematic code: (Response times vary from 2 -15 seconds)
	Using a bridge
	Using a temp accessory with 15 second delay in loop

 Well running code: (response times typically under 1 second and no flicker to LCD)
	Not using bridge
	Removed temp accessory, no longer exposed to HomeKit. Temp code now in main loop and provided to fan accessory thru global variable.  Tacky but it works.


As I cannot tell if the performance issues are related to multiple accessory loops, the drivers Im using to pull temperature and servo status  or the fact that a bridge is involved I will try a few different configurations using simulated services … My hope is to isolate the issue to some major component. 

Ultimately I need to determine where the issue is as this project will ultimately need to support 5 services  (fan, door, temp, battery and motion). It maybe that Im asking a bit much out of an ESP32.

**Comment by helo-head** [comment]
After a lot of troubleshooting I decided to just re-write the entire device loops based on the  "11 - ServiceOptions" example. Not sure how/why but it seems to be working just fine now. Rule number 26: Never argue with success.

**Comment by HomeSpan** [comment]
Thanks for the update - really glad to hear things seem to be working.  Let me know if any other issues surface and I'll see what I can do to help.
